### ISSUE TITLE
How to configure custom executor that limits number of threads, default we have cached executor

### ISSUE THREAD

**Issue by archit-harness** [grpc/grpc-java#10745]
Its recommended in the documentation that we should use the given - https://grpc.io/docs/guides/performance/ custom executor, but how do we configure this as its mentioned here - https://github.com/grpc/grpc-java/issues/7381 that due to many real scenarios we should stick to cached executor impl

**Comment by ejona86** [comment]
If you have no more clue than gRPC what the limits should be, yeah, you can't do much. But generally you have an idea for an upper bound to number of threads. You don't have to use a fixed-sized thread pool, but if you can, at least using a ThreadPoolExecutor with a maximum size which can help keep your system processing longer when overloaded. (ThreadPoolExecutor without any limit is the "cached thread pool.")

The number of needed threads is most significantly influenced by whether you do blocking within your service handlers.

**Comment by archit-harness** [comment]
We are doing blocking work in our servers, as the response ensures the work is completed by the server.
Also we are using netty grpc channels.

Questions mostly come around 
a) As you mentioned how to identify number of threads to the threadPoolExecutor for our service. (Should check our JFR or thread dump how many grpc threads are getting spawned during our load test?)

Reason -> In our JFR we can see there is a lot of context switching, thus we wanted to reduce and reuse threads to the fullest without unbounded number of threads.

https://grpc.io/docs/guides/performance/ -> also recommends the same so was further curious, as our live production code is working fine but we further want to scale and i think if we adopt this, this will improve server performance.


Is this how we are saying we should set our custom threadPool?
NettyChannelBuilder.forTarget(clientConfig.getTarget())
                    .overrideAuthority(authorityToUse)
                    .usePlaintext()
                    .executor(new ThreadPoolExecutor(...))
                    .maxInboundMessageSize(GrpcInProcessServer.GRPC_MAXIMUM_MESSAGE_SIZE)
                    .build();
                    
Also how do you recommend we should monitor our grpcs?

Basically if grpc average time is increased how to determine? is it grpc configuration and threads not available leading to it or our service is spending lot of time doing the work?

**Comment by ejona86** [comment]
> As you mentioned how to identify number of threads to the threadPoolExecutor for our service.

For async this is generally "number of cores" or the like. For blocking, it is probably the lesser of "maximum number of requests to process in parallel because of memory constraints" and "number of threads at which performance stops increasing." That second one would come into play if your application only blocks on a thread briefly; at a certain point you have enough concurrent RPCs that all the CPU cores are being used. Any threads beyond that point just adds memory/context switching.

> Is this how we are saying we should set our custom threadPool?
> .executor(new ThreadPoolExecutor(...))

Yes, that's how.

> Also how do you recommend we should monitor our grpcs?

There is grpc-census which uses OpenCensus. We're in the process of working on OpenTelemetry, but it isn't available yet.

**Comment by archit-harness** [comment]
> at a certain point you have enough concurrent RPCs that all the CPU cores are being used

How do you find that the threads are reached the optimal number? From our JFRs we can see there is high context switching, but we have various threadPool, so not sure how much grpc is contributing towards it, but how do i confirm this number?

Secondly, https://grpc.io/docs/guides/performance/ -> do you recommend to use pool of channels? and this pool we need to manage ourselves?

**Comment by ejona86** [comment]
> How do you find that the threads are reached the optimal number?

Worst-case, try a benchmark with a few different settings and look for the curve to form. At very low numbers, increasing number of threads improves throughput. At some point the gains slow/stop. There is often a pretty wide area where "anything in this range has pretty similar performance." You're trying to saturate your CPUs; so you are searching for the point you get close to 100% CPU utilization (unless you hit 100% memory utilization first!). It's commonly not worth it to get it "perfect" as it'd take too much effort and change the moment your workload or application changes, but just find something "reasonable."

> do you recommend to use pool of channels?

Generally not, mostly because people are too likely to do it when unnecessary. But sometimes it can really help. As a guideline, if you are doing 1 Gbps or more on a connection, adding more channels may make sense. See this old SO answer of mine: https://stackoverflow.com/a/58546457/4690866

**Comment by ejona86** [comment]
Seems like this is resolved. If not, comment, and it can be reopened.

**Comment by archit-harness** [comment]
Thanks @ejona86 for answering the questions promptly and resolving it
